### PR TITLE
Update fromStream for Projection, Query and ReadModelProjection

### PR DESCRIPTION
**Pull request by fjogeleit**
This change make it possible to query for a single aggregate with a SingleStream- and SimpleStreamStrategy behind.

### CODE REVIEW

File: src/Projection/InMemoryEventStoreQuery.php [hunk 1]
```diff
@@ -396,7 +414,7 @@ private function prepareStreamPositions(): void
         }
 
         // stream names given
-        foreach ($this->query['streams'] as $stream) {
+        foreach ($this->query['streams'] as $key => $stream) {
```
Comment (gquemener) [inline src/Projection/InMemoryEventStoreQuery.php]: `$key` is unused and should be ignored, shouldn't it?

File: src/Projection/Query.php [hunk 2]
```diff
@@ -29,6 +30,8 @@ public function fromStream(string $streamName): Query;
 
     public function fromStreams(string ...$streamNames): Query;
 
+    public function fromStreamWithMetadataMatcher(string $streamName, MetadataMatcher $metadataMatcher): Query;
```
Comment (gquemener) [inline src/Projection/Query.php]: This is a BC break and should be released in a new major version, shouldn't it?

File: src/Projection/InMemoryEventStoreQuery.php (hunk 1 above)
Comment (fjogeleit) [reply src/Projection/InMemoryEventStoreQuery.php]: yes, your right.

File: src/Projection/InMemoryEventStoreQuery.php [hunk 3]
```diff
@@ -79,6 +80,11 @@
      */
     private $triggerPcntlSignalDispatch;
 
+    /**
+     * @var MetadataMatcher
```
Comment (prolic) [inline src/Projection/InMemoryEventStoreQuery.php]: MetadataMatcher|null

File: src/Projection/Query.php (hunk 2 above)
Comment (prolic) [reply src/Projection/Query.php]: Exactly, my bad. I think this should work:
`public function fromStream(string $streamName, MetadataMatcher $metadataMatcher = null): Query;`

File: src/Projection/InMemoryEventStoreQuery.php (hunk 3 above)
Comment (fjogeleit) [reply src/Projection/InMemoryEventStoreQuery.php]: fixed

File: src/Projection/ReadModelProjector.php [hunk 4]
```diff
@@ -35,7 +36,7 @@
      */
     public function init(Closure $callback): ReadModelProjector;
 
-    public function fromStream(string $streamName): ReadModelProjector;
+    public function fromStream(string $streamName, MetadataMatcher $metadataMatcher = null): ReadModelProjector;
```
[the remark this answers is not in the document]
Comment (fjogeleit) [reply src/Projection/ReadModelProjector.php]: I removed it from the interfaces

**Comment by prolic** [comment]
Looks good so far. Before merging this, I'd like to see two things.
1) implementation for pdo event store
2) docs

**Comment by fjogeleit** [comment]
A PR for the PDO Event Store is already open:  https://github.com/prooph/pdo-event-store/pull/177

Docs will be updated asap

File: src/Projection/ReadModelProjector.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (gquemener) [reply src/Projection/ReadModelProjector.php]: Changing a method signature (even with optional parameters) is a BC break and all custom implementations will break after this change.

```
php > interface X
php > { 
php { public function foo(string $a, int $b = null): void
php { ;
php { }
php > class A implements X
php > {
php { public function foo(string $a): void
php { {
php { }
php { }

Fatal error: Declaration of A::foo(string $a): void must be compatible with X::foo(string $a, ?int $b = NULL): void in php shell code on line 1
```

A BC solution would have been to create a separate interface, which would have provided the `fromStreamWithMetadataMatcher` method OR to release this in a new major version.

**Comment by prolic** [comment]
The change is only in implementation, not in the interface itself.

**Comment by gquemener** [comment]
Oh right, I see.

Sorry, I got confused with the github discussion feed.

**Comment by prolic** [comment]
No problem

**Pull request merged**